### ISSUE TITLE
Openbsd 6.3 error: Protocol not available

### ISSUE THREAD

**Issue by raratiru** [celery/py-amqp#200]
* Openbsd-6.3
* Python-3.6
* Celery-4.2.0
* amqp-2.3.2

When starting Celery, I receive the error:

`[2018-07-18 00:53:50,510: ERROR/MainProcess] consumer: Cannot connect to amqp://...: [Errno 42] Protocol not available.`

**amqp-2.2.2** is the most recent version that works.

**Comment by neostimul** [comment]
The same situation with FreeBSD 11.2-RELEASE

**Comment by auvipy** [comment]
plz send a pr to fix the regression

**Comment by jasconius** [comment]
confirming this issue on FreeBSD. Using freebsd's package manager version of rabbitmq, but identical behavior... going to attempt to roll back to older rabbitmq to see if this addresses it

**Comment by clokep** [comment]
You likely need to modify the `KNOWN_TCP_OPTS` on FreeBSD, it seems to already have a bunch of platform specific changes in it.

**Comment by raratiru** [comment]
> You likely need to modify the `KNOWN_TCP_OPTS` on FreeBSD, it seems to already have a bunch of platform specific changes in it.

@clokep  [I just got confused! ](https://www.google.com/search?client=firefox-b&ei=W7O7W4ykIbGk_QaGwa6ADg&q=KNOWN_TCP_OPTS+freebsd&oq=KNOWN_TCP_OPTS+freebsd&gs_l=psy-ab.3...0.0.0.3176.0.0.0.0.0.0.0.0..0.0....0...1c..64.psy-ab..0.0.0....0.WnvE8AmJZCs) What kind of modification is requied?

**Comment by clokep** [comment]
@raratiru Check out the changes in #167, they touch this code that configures `KNOWN_TCP_OPTS`:

https://github.com/celery/py-amqp/blob/dc5b07c2320f932b2dd8342d0f4d578c095600cf/amqp/platform.py#L32-L57

Something is probably "wrong" with this code for FreeBSD. You could check what this code returns and attempt removing each option until you don't get this error. (Or compare what `KNOWN_TCP_OPTS` ends up as between 2.2.2 and the latest version.)

**Comment by raratiru** [comment]
OK, this is clear!

The author of #167 is referring to some failing tests however that [are unreachable now](https://codecov.io/gh/celery/py-amqp/pull/167?src=pr&el=tree).

> I've added some tests around this. Once again - I don't think I'm responsible for failing tests.

[git-bisect](https://git-scm.com/docs/git-bisect) could help in that case, if we had a failing test to run it against.

**Comment by raratiru** [comment]
> plz send a pr to fix the regression

@auvipy  There are two issues that I do not understand regarding the above prompt:

* Whom is it addressed to?
* How is this pr going to be tested, given that openbsd or freebsd or *bsd are not included in the [platforms](https://github.com/celery/py-amqp/blob/6cea54ed85318317af3f02c1da0e4bcdde6a4af3/amqp/platform.py#L41) tested against.

The problem seems to be deeper:

Are we sure that py-amqp is ready for a production bsd server?

**Comment by clokep** [comment]
@raratiru The above request for a PR was to anyone who cares about having BSD supported.

I'm unsure how it would be tested, especially since no CI is done on it. But likely people in this issue could be asked to test it.

**Comment by raratiru** [comment]
OK!

I will try to tackle with the `KNOWN_TCP_OPTS` differences between the two commits and report back the results.

**Comment by clokep** [comment]
As with the other if statements there, you'll likely need to check `sys.platform` or `platform.*` to separate changes from Linux. I don't have a BSD system around to play with to see what those appear as.

**Comment by raratiru** [comment]
I run git bisect which revealed something new:

The first bad commit is the following:

```
3faa7598536d8d8ccdf07722f5af8f2e44d9d0a1 is the first bad commit
commit 3faa7598536d8d8ccdf07722f5af8f2e44d9d0a1
Author: Dan Chowdhury <danthetaliban@gmail.com>
Date:   Mon Oct 23 16:19:24 2017 +0100

    #145: Cleanup TCP configurations across platforms and unified defaults into one dict (#167)
    
    * Cleanup TCP configurations across platforms and unified defaults into one dict
    
    * Added tests for platforms

:040000 040000 991d8008e0603228198a2bc10b426ef240f7c333 7f4d7192f4a9044cea33dd7d8e9fe09a6afb2b55 M      amqp
:040000 040000 c7741333bb6316c8581ee0a0ba9295ae17133c24 5fa8f07f779bd565812b31dcb3eb9db1a680aa0d M      t

```

This commit introduces the following error:
```
Traceback (most recent call last):
  File "/home/user/env/amqp/lib/python3.6/site-packages/celery/worker/worker.py", line 205, in start
    self.blueprint.start(self)
  File "/home/user/env/amqp/lib/python3.6/site-packages/celery/bootsteps.py", line 119, in start
    step.start(parent)
  File "/home/user/env/amqp/lib/python3.6/site-packages/celery/bootsteps.py", line 369, in start
    return self.obj.start()
  File "/home/user/env/amqp/lib/python3.6/site-packages/celery/worker/consumer/consumer.py", line 317, in start
    blueprint.start(self)
  File "/home/user/env/amqp/lib/python3.6/site-packages/celery/bootsteps.py", line 119, in start
    step.start(parent)
  File "/home/user/env/amqp/lib/python3.6/site-packages/celery/worker/consumer/connection.py", line 23, in start
    c.connection = c.connect()
  File "/home/user/env/amqp/lib/python3.6/site-packages/celery/worker/consumer/consumer.py", line 404, in connect
    conn = self.connection_for_read(heartbeat=self.amqheartbeat)
  File "/home/user/env/amqp/lib/python3.6/site-packages/celery/worker/consumer/consumer.py", line 411, in connection_for_read
    self.app.connection_for_read(heartbeat=heartbeat))
  File "/home/user/env/amqp/lib/python3.6/site-packages/celery/worker/consumer/consumer.py", line 435, in ensure_connected
    callback=maybe_shutdown,
  File "/home/user/env/amqp/lib/python3.6/site-packages/kombu/connection.py", line 405, in ensure_connection
    callback)
  File "/home/user/env/amqp/lib/python3.6/site-packages/kombu/utils/functional.py", line 332, in retry_over_time
    return fun(*args, **kwargs)
  File "/home/user/env/amqp/lib/python3.6/site-packages/kombu/connection.py", line 261, in connect
    return self.connection
  File "/home/user/env/amqp/lib/python3.6/site-packages/kombu/connection.py", line 802, in connection
    self._connection = self._establish_connection()
  File "/home/user/env/amqp/lib/python3.6/site-packages/kombu/connection.py", line 757, in _establish_connection
    conn = self.transport.establish_connection()
  File "/home/user/env/amqp/lib/python3.6/site-packages/kombu/transport/pyamqp.py", line 130, in establish_connection
    conn.connect()
  File "/home/user/sandbox/py-amqp/amqp/connection.py", line 282, in connect
    self.transport.connect()
  File "/home/user/sandbox/py-amqp/amqp/transport.py", line 85, in connect
    self.socket_settings, self.read_timeout, self.write_timeout,
  File "/home/user/sandbox/py-amqp/amqp/transport.py", line 137, in _init_socket
    self._set_socket_options(socket_settings)
  File "/home/user/sandbox/py-amqp/amqp/transport.py", line 176, in _set_socket_options
    tcp_opts = self._get_tcp_socket_defaults(self.sock)
  File "/home/user/sandbox/py-amqp/amqp/transport.py", line 172, in _get_tcp_socket_defaults
    tcp_opts[enum] = sock.getsockopt(SOL_TCP, opt)
TypeError: an integer is required (got type str)
```

Bisecting again, this commit:

```
73391a7d3dfc8b3f3dae918228ff482bb42e600b is the first bad commit
commit 73391a7d3dfc8b3f3dae918228ff482bb42e600b
Author: Matthias Erll <merll@users.noreply.github.com>
Date:   Mon Dec 11 10:50:36 2017 +0100

    Lookup by socket option name. (#173)

:040000 040000 1c5994f9481f6390bbc2031d8ad2f6d6adacecb7 9aeb8c5d62368fec597c1a4ced38d4601e729b18 M      amqp

```

Is the commit that inserts the current error:

`[2018-10-12 22:01:29,865: ERROR/MainProcess] consumer: Cannot connect to amqp://... [Errno 42] Protocol not available.`

**Comment by raratiru** [comment]
Here is the issue:

https://github.com/celery/py-amqp/blob/dc5b07c2320f932b2dd8342d0f4d578c095600cf/amqp/platform.py#L41-L44

`TCP_USER_TIMEOUT` should also be removed for openbsd.

In my case:
```
>>> sys.platform
'openbsd6'

```

**Comment by clokep** [comment]
I believe this is fixed by #211, but one of the reporters of this bug should confirm!

**Comment by raratiru** [comment]
Yes I can confirm, thank you! 

Looking forward to a new release so pip will grub the changes. 

I am closing this OpenBSD related issue.
